src/circuit/gadgets/constraints.rs

```diff
@@ -99,6 +120,61 @@ pub(crate) fn popcount<F: PrimeField, CS: ConstraintSystem<F>>(
     Ok(())
 }
 
+/// Adds a constraint to CS, enforcing that the addition of the allocated numbers in vector `v`
+/// is equal to `one`.
+///
+/// summation(v) = one
+#[allow(dead_code)]
+pub(crate) fn popcount_one<F: PrimeField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    v: &[Boolean],
+) -> Result<(), SynthesisError> {
+    let mut v_lc = LinearCombination::<F>::zero();
+    for b in v {
+        v_lc = add_to_lc::<F, CS>(b, v_lc, F::one())?;
+    }
+
+    // (summation(v)) * 1 = 1
+    cs.enforce(
+        || "popcount_one",
+        |_| v_lc,
+        |lc| lc + CS::one(),
+        |lc| lc + CS::one(),
+    );
+
+    Ok(())
+}
+
+/// Adds a constraint to CS, enforcing that the addition of the allocated numbers in vector `v`
+/// is equal to one if the premise is true.
+///
+/// summation(v) = one (if not dummy)
+#[allow(dead_code)]
+pub(crate) fn enforce_selector_with_premise<F: PrimeField, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    premise: &Boolean,
+    v: &[Boolean],
+) -> Result<(), SynthesisError> {
+    let mut v_lc = LinearCombination::<F>::zero();
+    for b in v {
+        v_lc = add_to_lc::<F, CS>(b, v_lc, F::one())?;
+    }
+    let mut nd_lc_b = LinearCombination::<F>::zero();
+    nd_lc_b = add_to_lc::<F, CS>(premise, nd_lc_b, F::one())?;
+    let mut nd_lc_c = LinearCombination::<F>::zero();
+    nd_lc_c = add_to_lc::<F, CS>(premise, nd_lc_c, F::one())?;
+
+    // (summation(v)) * premise = premise
+    cs.enforce(
+        || "enforce selector if not dummy",
+        |_| v_lc,
+        |_| nd_lc_b,
+        |_| nd_lc_c,
+    );
+
+    Ok(())
+}
```

Comment:
The way this is written obscures the meaning quite a bit. As much as possible, we need to try for clarity in the circuits. I would do something like this:
```
    let popcount = v
        .iter()
        .try_fold(LinearCombination::<F>::zero(), |acc, bit| {
            add_to_lc::<F, CS>(bit, acc, F::one())
        })?;

    let premise_b = premise.lc(CS::one(), F::one());
    let premise_c = premise_b.clone();

    // popcount(v) * premise = premise
    cs.enforce(
        || "enforce selector if not dummy",
        |_| popcount,
        |_| premise_b,
        |_| premise_c,
    );

    Ok(())
```

Reply:
Addressed in #404